src/jit/importer.cpp

```diff
@@ -6845,7 +6845,7 @@ var_types Compiler::impImportCall(OPCODE                  opcode,
         exactContextHnd                = callInfo->contextHandle;
         exactContextNeedsRuntimeLookup = callInfo->exactContextNeedsRuntimeLookup == TRUE;
 
-        // Recursive call is treaded as a loop to the begining of the method.
+        // Recursive call is treated as a loop to the begining of the method.
         if (methHnd == info.compMethodHnd)
```

Comment:
Is this the right test for recursion? (see `gtIsRecursiveCall` where we take care to properly handle sites inlinees).

Might be nice to ensure we have a test case with A->B->A call pattern, where B tailcalls A, A introduces a local struct, B gets inlined, and things will break if the local struct is not zeroed each time.

Reply:
It turns out we can't rely on BBF_BACKWARD_JUMP to block the struct newobj initialization optimization when tail recursion elimination is possible. It may be set too late, after newobj importing already happened. I added a test case that was failing with my previous changes. To fix that I modified fgMorphRecursiveFastTailCallIntoLoop to initialize temp structs with GC pointers on each iteration.

The test for recursion here is not correct as you indicated. Since this change no longer depends on it, I'd like to fix it in a separate PR. It will affect impMakeDiscretionaryInlineObservations so will likely result in diffs.

Reply:
Ok by me to fix this up later.